### DISCUSSION TITLE
How are AuxVariables calculated from material (qp) data?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#25099] [Q&A General] [answered]
Is this documented anywhere? Maybe in libMesh? For instance I have a MaterialRealAux. The documentation says it is "volume-averaged", but the material property is defined point-wise (ie no shape functions). Tried to reverse-engineer this by taking the qp values from my material (got them from a VPP) and doing a simple average, but that does not match the value of my constant monomial aux variable on that element. Also how does it work for a higher-order aux variable?

**Comment by GiudGiud** [comment]
You can compute the value of the material property at quadrature points then project back to some families. AuxKernels do that all the time, not just material properties aux

i don’t think we détail this in the documentation

**Reply by jessecarterMOOSE** [reply]
What's done for a simple constant monomial? Can you point to where this is done in the code? Looks like the Auxkernel returns the value, but I can't tell where that gets consumed.

**Reply by GiudGiud** [reply]
in AuxKernel.C it gets consumed. There's a special case for const monomial

**Reply by jessecarterMOOSE** [reply]
Is [this](https://github.com/idaholab/moose/blob/3ad745c5e36842feaeb601a35774c039560af827/framework/src/auxkernels/AuxKernel.C#L296-L309) the special case for const monomial? Looks like it's still doing some kind of quadrature. What rule does it use for monomials? I'm interested in both constant and first order.

**Reply by GiudGiud** [reply]
yeah it s using the quadrature to compute an average
1st order is right below. It attempts to do a least square-error minimizing projection

**Reply by jessecarterMOOSE** [reply]
Follow up: what gets saved in the Exodus file at nodes for the elemental variable? Does it do some kind of averaging of each element's projection to the node (for all elements that contain the node)? If so, is it just a straight arithmetic average?

**Reply by GiudGiud** [reply]
That s a different topic. 
for constant monomials, nothing I think. We would just do cell data. We don’t need to have nodal data. 

btw If you are trying to do projections between families you should use the ProjectionAux instead of relying on the auxkernel logic to do the right thing.

For higher order or different elemental families, it depends. Roy will know for each, which one do you care about?

**Reply by jessecarterMOOSE** [reply]
Right now I'm working backwards by taking the nodal values that are in the Exodus file for my material property and working backwards to try and figure out how they were calculated based on my qp. I'm really only interested up to first order, as well as the patch recovery method

**Reply by GiudGiud** [reply]
Can you put it in a nodal auxiliary variable, then use the ProjectionAux (which evaluates the nodal variable at qps then does the least square fitting), and see if that looks good for you?

**Reply by jessecarterMOOSE** [reply]
Oh can you do nodal aux variables from material properties? Didn't think about that. Still trying to figure out what my options are. Are there examples of that?

**Reply by GiudGiud** [reply]
oh yeah no actually you cant.
Not right now 

We have a plan for functor material properties to be nodal though

**Reply by jessecarterMOOSE** [reply]
So what's the "right" thing to do when you're interested in saving material property data to the exodus file?

**Reply by GiudGiud** [reply]
Figure out which shape you want and use the MaterialRealAux with an aux variable like that

**Reply by jessecarterMOOSE** [reply]
That's what I've been doing using constant and linear monomials. Just trying to figure out the caveats, i.e. how far off are the values when you extrapolate to nodes for Exodus, which presumably does a bunch of averaging from neighboring elements, then interpolate back across the element for visualization in Paraview.

**Reply by GiudGiud** [reply]
you can look here how we write elemental variables
https://libmesh.github.io/doxygen/classlibMesh_1_1ExodusII__IO__Helper.html#a32da5e71541d88d6e05a9df5547daf94

**Reply by jessecarterMOOSE** [reply]
Correct me if I'm wrong, but it looks like that method just writes elemental variables to file. The values at the nodes have already been computed elsewhere?

**Reply by GiudGiud** [reply]
~~Might have copied out the wrong one. There s a write_elemental in that file~~

Not sure. I m going to sidestep this and tag @roystgnr who will know

**Reply by roystgnr** [reply, marked as answer]
> @roystgnr who will know

I appreciate the optimism.  I'm afraid I never paid enough attention to this part of the code, though.  My attitude towards Exodus output quality has always been "eh, close enough for the eyeball" in the context of visualization and "what, are you mad, don't ever ever do that" in the context of restarts or postprocessing.

I think the best you can get from higher-order `MONOMIAL` in ExodusII is if you use a discontinuous output method, which then outputs them as nodal variables but which creates overlapping nodes and disconnected elements so you can still have multiple values at each node.  Each element's nodal values there are just the direct evaluation (of the limits) at that node.  And even in that context you still get (bi,tri) linear output for low-order `Elem` meshes and (bi,tri) quadratic for high-; with a few exceptions like EDGE4 and half-exceptions like TRI7 you're effectively throwing away p=3 and above terms.

If you try to output discontinuous variables via continuous nodal output, then we take the average of nodal evaluations from all elements sharing a node and plot that; we don't try to volume-weight it or anything.  (see "close enough" vs "are you mad" above)

**Reply by jessecarterMOOSE** [reply]
Thanks for the insight @roystgnr. Had a feeling there was some kind of averaging going on at the nodes or (at worst) a last-one-wins situation. Glad it's the former.

I'll give the discontinuous output a try.

**Reply by roystgnr** [reply]
Heh; you can thank Ben Kirk for that.  "Makes discontinuous solutions pretty when plotted at the nodes" is the 2003 commit message.  I'd personally have been satisfied with "last-one-wins" from a viz-only quality standpoint, but from a robustness standpoint it would have been unacceptable to get exodiffs from viz outputs of differently-partitioned runs.
